node/src/components/block_synchronizer.rs

```diff
@@ -1273,37 +1277,52 @@ impl<REv: ReactorEvent> Component<REv> for BlockSynchronizer {
                     // it's sent when we need to sync global states of an immediate switch block
                     // and its parent in order to check whether the validators have been
                     // changed by the upgrade
-                    BlockSynchronizerRequest::SyncGlobalStates(global_states, peers_to_ask) => {
-                        global_states
-                            .into_iter()
-                            .flat_map(move |(block_hash, global_state_hash)| {
-                                // only start syncing the state if we haven't started already
-                                if !self
-                                    .global_sync
-                                    .has_global_state_request(&global_state_hash)
-                                {
-                                    effect_builder
-                                        .sync_global_state(
-                                            block_hash,
-                                            global_state_hash,
-                                            peers_to_ask.clone().into_iter().collect(),
-                                        )
-                                        .ignore()
-                                } else {
-                                    Effects::new()
-                                }
-                            })
-                            .collect()
+                    BlockSynchronizerRequest::SyncGlobalStates(mut global_states) => {
+                        if let Some((block_hash, global_state_hash)) = global_states.pop() {
+                            let global_states_clone = global_states.clone();
+                            effect_builder
+                                .sync_global_state(block_hash, global_state_hash)
+                                .result(
+                                    move |_| {
+                                        Event::Request(BlockSynchronizerRequest::SyncGlobalStates(
+                                            global_states_clone,
+                                        ))
+                                    },
+                                    move |_| {
+                                        global_states.push((block_hash, global_state_hash));
+                                        Event::Request(BlockSynchronizerRequest::SyncGlobalStates(
+                                            global_states,
+                                        ))
+                                    },
```

Comment:
It might make sense, but if the hash that came from a block is wrong or otherwise un-syncable, we probably have bigger issues. And we used to have `fetch_retry_forever` functions in Fast Sync v1, too :wink:

Reply:
As agreed on the meeting, we'll merge as is.

If the solution proposed in this PR makes it to the final version of the node (instead of being replaced with rewritten global state synchronizer), we'll address that comment in a separate ticket.

Reply:
For the record: it's been agreed that "rewrite" will not be included in the release. The idea will be reconsidered in the future.